### ISSUE TITLE
Publish with retain doesn't retain

### ISSUE THREAD

**Issue by bldrdash** [mochi-co/mqtt#37]
Publish works fine except for honoring the retain flag.  Broker works fine with external publishing with retain.
Using master branch: 460f0ef6815ae482f85dc3b0b7ca01d789f70ad6

```golang
package main

import (
	"fmt"
	"os"
	"os/signal"
	"syscall"
	"time"

	mqtt "github.com/mochi-co/mqtt/server"
	"github.com/mochi-co/mqtt/server/listeners"
	"github.com/mochi-co/mqtt/server/listeners/auth"
)

func main() {
	sigs := make(chan os.Signal, 1)
	done := make(chan bool, 1)
	signal.Notify(sigs, syscall.SIGINT, syscall.SIGTERM)
	go func() {
		<-sigs
		done <- true
	}()

	server := mqtt.New()
	tcp := listeners.NewTCP("t1", ":1883")
	server.AddListener(tcp, &listeners.Config{Auth: new(auth.Allow)})

	// server.Serve doesn't block, so no need for goroutine, but for some
	// reason, we have to delay a LONG time for Serve's goroutines to start-up
	// > 1 second on a 5900X !!
	server.Serve()
	fmt.Println("\nServer is running")
	fmt.Println("Sleeping for 2 seconds")
	time.Sleep(2 * time.Second)

	// Publish with retain - doesn't work
	server.Publish("direct/publish", []byte("Published from broker"), true)
	fmt.Println("\nMessage published to \"direct/publish\" with retain")

	fmt.Println("\nTry running: mosquitto_sub -v -t \"direct/publish\"")

	<-done
}
```
```sh
go run main.go

# In another terminal window

# The following shows nothing
mosquitto_sub -v -t "direct/publish"

# But this works:
mosquitto_pub -t "mosquitto/publish" -m "from mosquitto_pub" -r
mosquitto_sub -v -t "mosquitto/publish"
```

Am I missing something?

**Comment by wind-c** [comment]
trie.go:
// RetainMessage saves a message payload to the end of a topic branch. Returns
// 1 if a retained message was added, 0 if there was no change, and -1 if the
// retained message was removed.
func (x *Index) RetainMessage(msg packets.Packet) int64 {
	var q int64

	x.mu.Lock()
	defer x.mu.Unlock()
	n := x.poperate(msg.TopicName)
	if len(msg.Payload) > 0 {
		**if n.Message.FixedHeader.Retain == false {
			q = 1
		}**
		n.Message = msg
	} else {
		if n.Message.FixedHeader.Retain == true {
			q = -1
		}
		x.unpoperate(msg.TopicName, "", true)
	}

	return q
}

The retain message is updated only when q=1，but this method always returns -1

**Comment by wind-c** [comment]
Error logic is as follows:
trie.go: RetainMessage method always return 0 if len(msg.payload) > 0
server.go:retainMessage method updates topic's retain message when q=1，and deletes stored retain message if q is not equal to 1.

**Comment by bldrdash** [comment]
#42 PR confirmed working for me

**Comment by mochi-co** [comment]
Fixed in v1.1.2 branch as per @bldrdash :)
